reduction/reduction.cu

```diff
@@ -0,0 +1,152 @@
+                      // power of 2
+  x--;
+  x |= x >> 1;
+  x |= x >> 2;
+  x |= x >> 4;
+  x |= x >> 8;
+  x |= x >> 16;
+  return ++x;
+}
+void GetNumBlocksAndThreads(unsigned int n, int& blocks,
+                            int& threads) {  // Assuming first add. Each thread
+                                             // brings in two values from mem
+  threads = (n > MAX_THREADS * 2) ? MAX_THREADS : nxtPow2((n + 1) / 2);
+  blocks = (n + threads * 2 - 1) / (threads * 2);
+}
+
+__device__ void WarpUnroll(volatile dtype* blk_mem, unsigned int tid) {
+  blk_mem[tid] += blk_mem[tid + 32];
+  blk_mem[tid] += blk_mem[tid + 16];
+  blk_mem[tid] += blk_mem[tid + 8];
+  blk_mem[tid] += blk_mem[tid + 4];
+  blk_mem[tid] += blk_mem[tid + 2];
+  blk_mem[tid] += blk_mem[tid + 1];
+}
+__global__ void ReductionKernel(dtype* i_data, unsigned int n, unsigned int N) {
+  __shared__ dtype blk_mem[MAX_THREADS];
+  unsigned int bid = blockIdx.y * gridDim.x + blockIdx.x;
+  unsigned int idx = bid * 2 * blockDim.x + threadIdx.x;
+
+  // bringing in main memory to block memory
+  if (idx >= N)
+    blk_mem[threadIdx.x] = 0.0;
+  else if (idx + blockDim.x >= N)
+    blk_mem[threadIdx.x] = i_data[idx];
+  else
+    blk_mem[threadIdx.x] =
+        i_data[idx] + i_data[idx + blockDim.x];  // Each thread brings in two
+                                                 // elemnts by adding them first
+  __syncthreads();
+
+  // Reducing block
+  for (unsigned int s = blockDim.x / 2; s > 32; s >>= 1) {
+    if (threadIdx.x < s) blk_mem[threadIdx.x] += blk_mem[threadIdx.x + s];
+    __syncthreads();
+  }
+  if (threadIdx.x < 32) WarpUnroll(blk_mem, threadIdx.x);
+
+  // writing back to main memory
+  if (threadIdx.x == 0) i_data[bid] = blk_mem[0];
+}
+
+dtype Reduction(dtype* i_data_h, unsigned int N) {
+  dtype* i_data_d;
+  // Move data to GPU
+  cudaMalloc(&i_data_d, sizeof(dtype) * N);
+  cudaMemcpy(i_data_d, i_data_h, sizeof(dtype) * N, cudaMemcpyHostToDevice);
+  // Reduce
+  unsigned int n = N;
+  int threads = 0;
+  int blocks = 0;
+  while (n > 1) {
+    GetNumBlocksAndThreads(n, blocks, threads);
+    // Kernel launch
+    dim3 grid_block(16, (blocks + 16 - 1) / 16);
+    dim3 thread_block(threads, 1);
+    ReductionKernel<<<grid_block, thread_block>>>(i_data_d, n, N);
+    n = blocks;  // new problem size
+  }
+  // Move result back
+  cudaMemcpy(i_data_h, i_data_d, sizeof(dtype) * 1, cudaMemcpyDeviceToHost);
+  // Free
+  cudaFree(i_data_d);
+
+  return i_data_h[0];
+}
+
+dtype CPUReduction(dtype* i_data, unsigned int N) {
+  dtype sum = 0;
+  for (int i = 0; i < N; i++) {
+    sum += i_data[i];
+  }
+  return sum;
+}
+
+int main(int argc, char** argv) {
+  unsigned int N;
+  dtype* i_data_h;
+
+  // Procure arguments from user
+  if (argc > 1) {
+    N = (unsigned int)atoi(argv[1]);
+    printf("N = %u\n", N);
+  } else {
+    N = N_;
+    printf("N = %u\n", N);
+  }
+
+  // Prepare random input
+  i_data_h = (dtype*)malloc(sizeof(dtype) * N);
```

Comment:
Mallor/free and new/delete should almost never appear in a modern C++, outside of library implementations. 
std::vector<> would be fine.